hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java

```diff
@@ -362,33 +362,33 @@ public RegisteredCommand register(
           .build();
     }
 
+    InetAddress dnAddress = Server.getRemoteIp();
+    if (dnAddress != null) {
+      // Mostly called inside an RPC, update ip
+      datanodeDetails.setIpAddress(dnAddress.getHostAddress());
+    }
+
+    String dnsName;
+    String networkLocation;
+    datanodeDetails.setNetworkName(datanodeDetails.getUuidString());
+    if (useHostname) {
+      dnsName = datanodeDetails.getHostName();
+    } else {
+      dnsName = datanodeDetails.getIpAddress();
+    }
+    networkLocation = nodeResolve(dnsName);
+    if (networkLocation != null) {
+      datanodeDetails.setNetworkLocation(networkLocation);
+    }
+
     if (!isNodeRegistered(datanodeDetails)) {
-      InetAddress dnAddress = Server.getRemoteIp();
-      if (dnAddress != null) {
-        // Mostly called inside an RPC, update ip and peer hostname
-        datanodeDetails.setHostName(dnAddress.getHostName());
```

Comment:
I delete this line, because these days, when I tested it, I found sometimes dnAddress.getHostName() returns IP instead of hostName, which makes the datanode restarting not work. Please let me know if it is OK to delete this line.

Reply:
I am looking into the verification failures. How can I run a .robot test case locally? For example, I just want to run `/opt/hadoop/smoketest/recon/recon-api.robot` this test case. @adoroszlai

Reply:
@sokui @adoroszlai Sorry to say that I'm nervous about removing the call to  setHostName().

I just took around and it seems to get used in many places.  I've included some below:
https://github.com/apache/ozone/blob/f57a0193c114ac2f17dc255ec1a052dcf12781b8/hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientManager.java#L259

https://github.com/apache/ozone/blob/f57a0193c114ac2f17dc255ec1a052dcf12781b8/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/DatanodeChunkGenerator.java#L171

https://github.com/apache/ozone/blob/f57a0193c114ac2f17dc255ec1a052dcf12781b8/hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java#L576

Why does restart not work when it returns the IP string instead of the host string?

Reply:
Hi @GeorgeJahad ,

To not change the old code path, I added the if condition: when `useHostname` is true, we will not setHostname, but when it is false (old code), we keep the old logic which set the hostname. There are two things I want to explain:

1. When datanode fist register with scm, the `datanodeDetails` already contains hostName. Here the code we are talking about is just to reset the `datanodeDetails.hostName`. So the code you listed above won't return null if we remove this line of code `datanodeDetails.setHostName(dnAddress.getHostName());`.

2. Why it doesn't work when we reset `datanodeDetails.hostName` when `useHostname` is true? this is because in k8s, when datanode first registered with scm, `dnAddress.getHostName()` may return IP instead of hostName (maybe because of k8s DNS lookup service delay, I am not exactly sure). this will result in the IP instead of hostName is used in datanode Ratis communication for Pipelines. When datanode gets restarted with different IP, then the Ratis communication with old IP throws the HostNotFoundException. But if we remove this line, then we are sure that  the `datanodeDetails.hostName` always contains the hostname instead of the IP. So it won't have the Ratis communication problem.

This is the whole story. That's why now I keep the old code path same, but if `useHostname` is true, we won't do `datanodeDetails.setHostName(dnAddress.getHostName());` in the register process. Please let me know if it makes sense to you.

Reply:
>  When datanode fist register with scm, the datanodeDetails already contains hostName. Here the code we are talking about is just to reset the datanodeDetails.hostName.

If you are sure this is true, then I'm fine with the change.

> To not change the old code path, I added the if condition: when useHostname is true, we will not setHostname, but when it is false (old code),

I'm confused about this statement.  The old code path is when "(!isNodeRegistered(datanodeDetails))" is true, isn't it? not when useHostname is true?  what am I missing?

Reply:
Old codepath means current `master`, before this PR.

Reply:
What I was saying is that even we delete this line `datanodeDetails.setHostName(dnAddress.getHostName());`, it should be still fine because when datanode register with scm, the datanodeDetails already have the hostName info. But to be conservative, I just use the above logic to make sure when `DFS_DATANODE_USE_DN_HOSTNAME` is not enabled, the code is the exactly same as before.